crates/aquascope/src/analysis/permissions/flow.rs

```diff
@@ -0,0 +1,474 @@
+      // borrows don't conflict behind shared references.
+      root_place.projection = TyCtxtConsts::DEREF_PROJECTION;
+      (true, true)
+    } else {
+      (false, ctxt.locals_are_invalidated_at_exit)
+    };
+
+  if !will_be_dropped {
+    log::debug!(
+      "place_is_invalidated_at_exit({:?}) - won't be dropped",
+      place
+    );
+    return false;
+  }
+
+  let sd = if might_be_alive { Deep } else { Shallow(None) };
+
+  places_conflict::borrow_conflicts_with_place(
+    tcx,
+    body,
+    place,
+    borrow.kind,
+    root_place,
+    sd,
+    places_conflict::PlaceConflictBias::Overlap,
+  )
+}
+
+// ------------------
+// Entry
+
+pub fn compute_flows(ctxt: &mut PermissionsCtxt) {
+  let timer = Instant::now();
+  let tcx = ctxt.tcx;
+  let body = &ctxt.body_with_facts.body;
+
+  // Compute the constraint graph with all regions.
+  let constraints = ctxt
+    .polonius_input_facts
+    .subset_base
+    .iter()
+    .map(|&(o1, o2, _)| (o1, o2))
+    .collect::<Vec<_>>();
+
+  let vertices = flatten_tuples(&constraints).collect::<HashSet<_>>();
+
+  log::debug!("Constraints---\n{:#?}", constraints);
+
+  // Graph of constraints that need to be satisfied. This shows
+  // us how data flows from one region into another.
+  let constraint_graph = VecGraph::new(count_nodes(&constraints), constraints);
+
+  let scc_constraints = Sccs::<Origin, SccIdx>::new(&constraint_graph);
+  let num_sccs = scc_constraints.num_sccs();
+
+  log::debug!("There are a total of {num_sccs} SCCs");
+
+  // Regions that only occur in the return type are not
+  // included in the abstract placeholders set. Example:
+  // ```rust,ignore
+  // fn mk_string() -> &'a String {
+  //   let s = String::from("s");
+  //   &s
+  // }
+  // ```
+  // The `'a`, does not appear in the placeholders set.
+  let placeholders = ctxt
+    .polonius_input_facts
+    .placeholder
+    .iter()
+    .filter_map(|&(p, _)| vertices.contains(&p).then_some(p))
+    .chain(body.regions_in_return().map(|rg| rg.to_region_vid()))
+    .map(|p| scc_constraints.scc(p))
+    .collect::<Vec<_>>();
+
+  log::debug!("placeholders: {placeholders:#?}");
+
+  // We filter the placeholders because the `known_placeholder_subset` contains a
+  // top and bottom of the abstract lattice, but we only care about those
+  // that actually appear in the body.
+  let placeholder_edges = ctxt
+    .polonius_input_facts
+    .known_placeholder_subset
+    .iter()
+    .filter(|(f, t)| vertices.contains(f) && vertices.contains(t))
+    .map(|&(f, t)| (scc_constraints.scc(f), scc_constraints.scc(t)))
+    .chain(placeholders.iter().map(|&o| (o, o)))
+    .collect::<Vec<_>>();
+
+  log::debug!("placeholder_edges: {placeholder_edges:#?}");
+
+  // Allowed flows between abstract regions.
+  let known_flows_graph = VecGraph::new(num_sccs, placeholder_edges);
+
+  // Compute the flow facts between abstract regions.
+  let mut known_flows = BitMatrix::new(num_sccs, num_sccs);
```

Comment:
We should probably revisit some of the terminology.  I think `known` is not the best adjective here. Perhaps `user_specified_flows`?